GeneratorInterface/GenFilters/plugins/MCLongLivedParticles.cc

```diff
@@ -0,0 +1,95 @@
+
+MCLongLivedParticles::MCLongLivedParticles(const edm::ParameterSet& iConfig)
+    : moduleLabel_(iConfig.getParameter<std::string>("hepMCProductTag")),
+      hepMCProductTag_(
+          edm::InputTag(iConfig.getUntrackedParameter(moduleLabel_, std::string("generator")), "unsmeared")),
+      token_(consumes<edm::HepMCProduct>(hepMCProductTag_)),
+      particleIDs(iConfig.getParameter<std::vector<int>>("ParticleIDs")),
+      theUpperCut_(iConfig.getParameter<double>("LengMax")),
+      theLowerCut_(iConfig.getParameter<double>("LengMin")) {}
+
+void MCLongLivedParticles::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+  desc.add<std::string>("hepMCProductTag", "moduleLabel");
+  desc.add<std::vector<int>>("ParticleIDs", std::vector<int>{0});
+  desc.add<double>("LengMax", -1.);
+  desc.add<double>("LengMin", -1.);
+  descriptions.add("selectsDisplacement", desc);
+}
+
+// ------------ method called to skim the data  ------------
+bool MCLongLivedParticles::filter(edm::StreamID, edm::Event& iEvent, const edm::EventSetup& iSetup) const {
+  using namespace edm;
+
+  Handle<HepMCProduct> evt;
+
+  iEvent.getByToken(token_, evt);
+
+  bool pass = false;
+  bool matchedID = true;
+
+  float theUpperCut2 = theUpperCut_ * theUpperCut_;
+  float theLowerCut2 = theLowerCut_ * theLowerCut_;
+
+  if (particleIDs.at(0) != 0)
+    matchedID = false;
+
+  const HepMC::GenEvent* generated_event = evt->GetEvent();
+  HepMC::GenEvent::particle_const_iterator p;
+
+  for (p = generated_event->particles_begin(); p != generated_event->particles_end(); p++) {
+    //if a list of pdgId is provided, loop only on particles with those pdgId
+
+    for (unsigned int idx = 0; idx < particleIDs.size(); idx++) {
+      if (abs((*p)->pdg_id()) == abs(particleIDs.at(idx))) {  //compares absolute values of pdgIds
+        matchedID = true;
+        break;
+      }
+    }
+
+    if (matchedID) {
+      if (theLowerCut_ <= 0. && theUpperCut_ <= 0.) {
+        pass = true;
+        break;
+      }
+      if (((*p)->production_vertex() != nullptr) && ((*p)->end_vertex() != nullptr)) {
+        float dist2 = (((*p)->production_vertex())->position().x() - ((*p)->end_vertex())->position().x()) *
+                          (((*p)->production_vertex())->position().x() - ((*p)->end_vertex())->position().x()) +
+                      (((*p)->production_vertex())->position().y() - ((*p)->end_vertex())->position().y()) *
+                          (((*p)->production_vertex())->position().y() - ((*p)->end_vertex())->position().y());
+
+        if ((dist2 >= theLowerCut2 || theLowerCut_ <= 0.) &&
+            (dist2 < theUpperCut2 ||
+             theUpperCut_ <= 0.)) {  //lower cut can be also 0 - prompt particle needs to be accepted in that case
+          pass = true;
+          break;
+        }
+      }
+      if (((*p)->production_vertex() == nullptr) && (!((*p)->end_vertex() != nullptr))) {
```

Comment:
```suggestion
```
This can remain in the same `if block` as above
(But perhaps you wanted to check only the production vertex above, and the end vertex here... Please check)

Reply:
Yes, sorry, I tested the code with and without the check, it is needed to prevent from running on particles missing the prod vertex